validate/validate.go

```diff
@@ -0,0 +1,215 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package validate provides the library functions to validate a TDX quote
+package validate
+
+import (
+	"bytes"
+	"encoding/hex"
+	"fmt"
+
+	"github.com/google/go-tdx-guest/abi"
+	cpb "github.com/google/go-tdx-guest/proto/check"
+	pb "github.com/google/go-tdx-guest/proto/tdx"
+	vr "github.com/google/go-tdx-guest/verify"
+	"go.uber.org/multierr"
+)
+
+// Options represents validation options for a TDX attestation Quote.
+type Options struct {
+	HeaderOptions      HeaderOptions
+	TdQuoteBodyOptions TdQuoteBodyOptions
+}
+
+// HeaderOptions represents validation options for a TDX attestation Quote Header.
+type HeaderOptions struct {
+	// MinimumQeSvn is the minimum QE security version number. Must be nil or 2 bytes long. Not checked if nil.
+	MinimumQeSvn []byte
+	// MinimumPceSvn is the minimum PCE security version number. Must be nil or 2bytes long. Not checked if nil.
+	MinimumPceSvn []byte
+	// QeVendorID is the expected QE_VENDOR_ID field. Must be nil or 16 bytes long. Not checked if nil.
+	QeVendorID []byte
+}
+
+// TdQuoteBodyOptions represents validation options for a TDX attestation Quote's TD Quote body.
+type TdQuoteBodyOptions struct {
+	// MinimumTeeTcbSvn is the minimum TEE_TCB security version number. Must be nil or 16 bytes long. Not checked if nil.
+	MinimumTeeTcbSvn []byte
+	// MrSeam is the expected MR_SEAM field. Must be nil or 48 bytes long. Not checked if nil.
+	MrSeam []byte
+	// TdAttributes is the expected TD_ATTRIBUTES field. Must be nil or 8 bytes long. Not checked if nil.
```

Comment:
TdAttributes seems like it should be governed by a min/max range rather than an explicit single configuration. There's an attributes mask for the reason of having certain attributes be allowed (0 <= flag <= 1), but not necessarily required (1 <= flag <= 1). Conversely that also means there are some attributes that are not allowed (0 <= flag <= 0).

Reply:
Added a validateTdAttributes func for the TdAttributes field.